packages/trackerless-network/src/logic/RandomGraphNode.ts

```diff
@@ -0,0 +1,309 @@
+    nearbyContactPoolIdAdded: () => void
+}
+
+export interface StrictRandomGraphNodeConfig {
+    randomGraphId: string
+    layer1: DhtNode
+    P2PTransport: ITransport
+    connectionLocker: ConnectionLocker
+    ownPeerDescriptor: PeerDescriptor
+    peerViewSize: number
+    nearbyContactPool: PeerList
+    randomContactPool: PeerList
+    targetNeighbors: PeerList
+    handshaker: IHandshaker
+    neighborFinder: INeighborFinder
+    neighborUpdateManager: INeighborUpdateManager
+    propagation: Propagation
+    rpcCommunicator: ListeningRpcCommunicator
+    numOfTargetNeighbors: number
+    maxNumberOfContacts: number
+    minPropagationTargets: number
+    nodeName: string
+    acceptProxyConnections: boolean
+    neighborUpdateInterval: number
+    proxyConnectionServer?: ProxyStreamConnectionServer
+}
+
+const logger = new Logger(module)
+
+export class RandomGraphNode extends EventEmitter<Events> implements IStreamNode {
+    private stopped = false
+    private started = false
+    private readonly duplicateDetectors: Map<string, DuplicateMessageDetector>
+    private readonly abortController: AbortController
+    private config: StrictRandomGraphNodeConfig
+    private readonly server: INetworkRpc
+
+    constructor(config: StrictRandomGraphNodeConfig) {
+        super()
+        this.config = config
+        this.duplicateDetectors = new Map()
+        this.abortController = new AbortController()
+        this.server = new StreamNodeServer({
+            ownPeerDescriptor: this.config.ownPeerDescriptor,
+            randomGraphId: this.config.randomGraphId,
+            rpcCommunicator: this.config.rpcCommunicator,
+            markAndCheckDuplicate: (msg: MessageRef, prev?: MessageRef) => this.markAndCheckDuplicate(msg, prev),
+            broadcast: (message: StreamMessage, previousPeer?: string) => this.broadcast(message, previousPeer),
+            onLeaveNotice: (notice: LeaveStreamNotice) => {
+                const senderId = notice.senderId
+                const contact = this.config.nearbyContactPool.getNeighborWithId(senderId)
+                || this.config.randomContactPool.getNeighborWithId(senderId)
+                || this.config.targetNeighbors.getNeighborWithId(senderId)
+                || this.config.proxyConnectionServer?.getConnection(senderId as PeerIDKey)?.remote
+                // TODO: check integrity of notifier?
+                if (contact) {
+                    this.config.layer1.removeContact(contact.getPeerDescriptor(), true)
+                    this.config.targetNeighbors.remove(contact.getPeerDescriptor())
+                    this.config.nearbyContactPool.remove(contact.getPeerDescriptor())
+                    this.config.connectionLocker.unlockConnection(contact.getPeerDescriptor(), this.config.randomGraphId)
+                    this.config.neighborFinder.start([senderId])
+                    this.config.proxyConnectionServer?.removeConnection(senderId as PeerIDKey)
```

Comment:
Are all these actions needed for all contact nodes? I.e. if both `neighborWithId` methods return `undefined`, would it be enough just to call `this.config.proxyConnectionServer?.removeConnection(senderId as PeerIDKey)`?

Reply:
If a node leaves a stream it's best to remove it from all known structs that could hold data of it.